qr_scanner/__init__.py

```diff
@@ -0,0 +1,216 @@
+        """
+        pass
+
+    def load_function(self, name, argtypes=[], restype=None):
+        """
+        Load a given native function and create a field of the same name.,
+        """
+        function = getattr(self.lib, name)
+        function.argtypes = list(argtypes)
+        function.restype = restype
+        setattr(self, name, function)
+
+    def load_functions(self, functions):
+        """
+        Load all given native functions and create fields of the same name.
+        The method expect an iterable of function descriptions, where function
+        description is a triplet (name, argtypes, restype).
+        """
+        for function in functions:
+            self.load_function(*function)
+
+
+class QRScannerLibrary(Library):
+    """
+    Loader for libqr_scanner.
+    """
+
+    library = 'qr_scanner'
+
+    LOG_LEVEL_DEBUG = 1
+    LOG_LEVEL_INFO = 2
+    LOG_LEVEL_WARN = 3
+    LOG_LEVEL_ERROR = 4
+
+    LOG_CALLBACK = CFUNCTYPE(None, c_void_p, c_char_p, c_int, c_int, c_char_p)
+
+    def __init__(self):
+        super().__init__()
+
+        def log_callback(opaque, file, line, level, msg):
+            if level < self.LOG_LEVEL_DEBUG:
+                return
+            file = file.decode('utf-8')
+            msg = msg.decode('utf-8')
+            logger.debug(msg, extra={
+                'file': '{}:{}'.format(file, line),
+            })
+
+        self.__log_callback = self.LOG_CALLBACK(log_callback)
+
+        self.qrs__set_log_callback(self.__log_callback, None)
+
+    def load_symbols(self):
+        self.load_functions((
+            ('qrs__set_log_callback', [self.LOG_CALLBACK, c_void_p]),
+
+            ('qrs__scanner_config__new', [], c_void_p),
+            ('qrs__scanner_config__free', [c_void_p]),
+            ('qrs__scanner_config__key_frames_only', [c_void_p, c_int]),
+
+            ('qrs__scanner__scan_http_stream', [c_char_p, c_uint64, c_void_p], c_void_p),
+            ('qrs__scanner__free', [c_void_p]),
+            ('qrs__scanner__next_qr_code', [c_void_p], c_void_p),
+
+            ('qrs__qr_code__free', [c_void_p]),
+            ('qrs__qr_code__get_data', [c_void_p], POINTER(c_uint8)),
+            ('qrs__qr_code__get_data_size', [c_void_p], c_size_t),
+        ))
+
+
+lib = QRScannerLibrary()
+
+
+class NativeObject:
+    """
+    Abstraction over a native object. The memory will be released automatically
+    by garbage collector if a free_func is given.
+    """
+
+    def __init__(self, raw_ptr: c_void_p, free_func=None):
+        self.__raw_ptr = raw_ptr
+        self.__free_func = free_func
+        self.__free_lock = None if free_func is None else threading.Lock()
+
+    def __del__(self):
+        if self.__free_func is None:
+            return
+        if self.__raw_ptr is None:
+            return
+
+        with self.__free_lock:
+            if self.__raw_ptr is not None:
+                self.__free_func(self.__raw_ptr)
+            self.__raw_ptr = None
```

Comment:
It's weird but it really is necessary. I would also prefer not having it there.

Originally, we had it without the lock in the stream router lib but my-angelcam was sometimes crashing on double-free when the app was terminating. It seems that cPython can call the `__del__` method multiple times from different threads at the same time. The weird thing is that the GIL is there to prevent exactly this type of race conditions. :man_shrugging:

Reply:
Well there is a note in https://docs.python.org/3/reference/datamodel.html#object.__del__ 
about resurrecting objects, but that doesn't explain such behavior.

Anyways thanks for the explanation.